pulp_python/app/tasks/upload.py

```diff
@@ -28,53 +30,81 @@
+
+
+def create_content(artifact_sha256, filename):
+    """
+    Creates PythonPackageContent from artifact.
+
+    Args:
+        artifact_sha256: validated artifact
+        filename: file name
+    Returns:
+        queryset of the new created content
+    """
     # iterate through extensions since splitext does not support things like .tar.gz
-    for ext, packagetype in DIST_EXTENSIONS.items():
-        if filename.endswith(ext):
-            # Copy file to a temp directory under the user provided filename, we do this
-            # because pkginfo validates that the filename has a valid extension before
-            # reading it
-            with tempfile.TemporaryDirectory() as td:
-                temp_path = os.path.join(td, filename)
-                artifact = Artifact.objects.get(pk=artifact_pk)
-                shutil.copy2(artifact.file.path, temp_path)
-                metadata = DIST_TYPES[packagetype](temp_path)
-                metadata.packagetype = packagetype
-                break
-    else:
-        raise serializers.ValidationError(_(
-            "Extension on {} is not a valid python extension "
-            "(.whl, .exe, .egg, .tar.gz, .tar.bz2, .zip)").format(filename)
-        )
+    extensions = list(DIST_EXTENSIONS.keys())
+    pkg_type_index = [filename.endswith(ext) for ext in extensions].index(True)
+    packagetype = DIST_EXTENSIONS[extensions[pkg_type_index]]
+    # Copy file to a temp directory under the user provided filename, we do this
+    # because pkginfo validates that the filename has a valid extension before
+    # reading it
+    with tempfile.TemporaryDirectory() as td:
+        temp_path = os.path.join(td, filename)
+        artifact = Artifact.objects.get(sha256=artifact_sha256)
+        shutil.copy2(artifact.file.path, temp_path)
+        metadata = DIST_TYPES[packagetype](temp_path)
+        metadata.packagetype = packagetype
+
     data = parse_project_metadata(vars(metadata))
-    data['classifiers'] = [{'name': classifier} for classifier in metadata.classifiers]
     data['packagetype'] = metadata.packagetype
     data['version'] = metadata.version
     data['filename'] = filename
-    data['_relative_path'] = filename
-
-    new_content = PythonPackageContent.objects.create(
-        filename=filename,
-        packagetype=metadata.packagetype,
-        name=data['classifiers'],
-        version=data['version']
-    )
+    data['sha256'] = artifact.sha256
 
-    queryset = PythonPackageContent.objects.filter(pk=new_content.pk)
-
-    if repository_pk:
-        repository = PythonRepository.objects.get(pk=repository_pk)
-        with repository.new_version() as new_version:
-            new_version.add_content(queryset)
+    @transaction.atomic()
+    def create():
+        content = PythonPackageContent.objects.create(**data)
+        ContentArtifact.objects.create(
+            artifact=artifact, content=content, relative_path=filename
+        )
+        return content
```

Comment:
This should probably be wrapped in a transaction, if it isn't already (and maybe it is)

Reply:
It is.